server/src/test/java/org/elasticsearch/gateway/GatewayServiceTests.java

```diff
@@ -88,4 +141,355 @@ public void testRecoverStateUpdateTask() throws Exception {
+        final Settings.Builder settingsBuilder = Settings.builder();
+        final int expectedNumberOfDataNodes = randomIntBetween(4, 6);
+        final boolean hasRecoverAfterTime = randomBoolean();
+        if (hasRecoverAfterTime) {
+            settingsBuilder.put(RECOVER_AFTER_TIME_SETTING.getKey(), TimeValue.timeValueMinutes(10));
+        } else {
+            settingsBuilder.put(EXPECTED_DATA_NODES_SETTING.getKey(), expectedNumberOfDataNodes);
+        }
+        final int recoverAfterNodes = expectedNumberOfDataNodes - 1;
+        settingsBuilder.put(RECOVER_AFTER_DATA_NODES_SETTING.getKey(), recoverAfterNodes);
+
+        final ClusterState initialState = buildClusterState(1, 1);
+        final GatewayService gatewayService = createService(settingsBuilder, initialState);
+        assertClusterStateBlocks(true);
+
+        // Not recover because recoverAfterDataNodes not met
+        final long initialTimeInMillis = deterministicTaskQueue.getCurrentTimeMillis();
+        final var setDataNodeCountTaskQueue = createSetDataNodeCountTaskQueue(clusterService);
+        setDataNodeCountTaskQueue.submitTask(randomAlphaOfLength(5), new SetDataNodeCountTask(recoverAfterNodes - 1), null);
+        deterministicTaskQueue.runAllTasksInTimeOrder();
+        assertThat(deterministicTaskQueue.getCurrentTimeMillis(), equalTo(initialTimeInMillis));
+        assertClusterStateBlocks(true);
+        final var pendingStateRecoveryOfInitialTerm = gatewayService.currentPendingStateRecovery;
+        assertThat(pendingStateRecoveryOfInitialTerm, notNullValue());
+
+        // The 1st scheduled recovery when recoverAfterDataNodes is met
+        setDataNodeCountTaskQueue.submitTask(randomAlphaOfLength(5), new SetDataNodeCountTask(recoverAfterNodes), null);
+        deterministicTaskQueue.runAllRunnableTasks();
+        assertThat(deterministicTaskQueue.hasDeferredTasks(), is(true));
+        assertThat(gatewayService.currentPendingStateRecovery, sameInstance(pendingStateRecoveryOfInitialTerm));
+
+        // The 1st schedule is cancelled when data nodes drop below recoverAfterDataNodes
+        final TimeValue elapsed = TimeValue.timeValueMinutes(1);
+        deterministicTaskQueue.scheduleAt(
+            initialTimeInMillis + elapsed.millis(),
+            () -> setDataNodeCountTaskQueue.submitTask(randomAlphaOfLength(5), new SetDataNodeCountTask(recoverAfterNodes - 1), null)
+        );
+        deterministicTaskQueue.advanceTime();
+        deterministicTaskQueue.runAllRunnableTasks();
+
+        // The 2nd scheduled recovery when data nodes are above recoverAfterDataNodes again
+        deterministicTaskQueue.scheduleAt(
+            initialTimeInMillis + elapsed.millis() * 2,
+            () -> setDataNodeCountTaskQueue.submitTask(randomAlphaOfLength(5), new SetDataNodeCountTask(recoverAfterNodes), null)
+        );
+        deterministicTaskQueue.advanceTime();
+        deterministicTaskQueue.runAllRunnableTasks();
```

Comment:
These `advanceTime` then `runAllRunnableTasks()` can be simplified if we have some method like `runToTime(long)` that runs all tasks up to the specified time. I can have a follow-up PR to add this method if you think it is useful.

Reply:
Yep I can see value in that, please do.